identifier/identifier.go

```diff
@@ -75,3 +80,70 @@ func NewIP(ip netip.Addr) ACMEIdentifier {
 		Value: ip.String(),
 	}
 }
+
+// FromCert extracts the Subject Alternative Names from a certificate, and
+// returns a slice of ACMEIdentifiers and an error. It does not extract the
+// Subject Common Name.
+//
+// FromCSR is similar but handles CSRs, and is kept separate so that it's always
+// clear we are handling an untrusted CSR.
+func FromCert(cert *x509.Certificate) []ACMEIdentifier {
+	var sans []ACMEIdentifier
+	for _, name := range cert.DNSNames {
+		sans = append(sans, NewDNS(name))
+	}
+
+	for _, ip := range cert.IPAddresses {
+		sans = append(sans, ACMEIdentifier{
+			Type:  TypeIP,
+			Value: ip.String(),
+		})
+	}
+
+	return Normalize(sans)
+}
+
+// FromCSR extracts the Subject Common Name and Subject Alternative Names from a
+// CSR, and returns a slice of ACMEIdentifiers and an error.
+//
+// FromCert is similar but handles certs, and is kept separate so that it's
+// always clear we are handling an untrusted CSR.
+func FromCSR(csr *x509.CertificateRequest) []ACMEIdentifier {
+	var sans []ACMEIdentifier
+	for _, name := range csr.DNSNames {
+		sans = append(sans, NewDNS(name))
+	}
+	if csr.Subject.CommonName != "" {
+		// Boulder won't generate certificates with a CN that's not also present
+		// in the SANs, but such a certificate is possible. If appended, this is
+		// deduplicated later with Normalize(). We assume the CN is a DNSName,
+		// because CNs are untyped strings without metadata, and we will never
+		// configure a Boulder profile to issue a certificate that contains both
+		// an IP address identifier and a CN.
+		sans = append(sans, NewDNS(csr.Subject.CommonName))
+	}
+
+	for _, ip := range csr.IPAddresses {
+		sans = append(sans, ACMEIdentifier{
+			Type:  TypeIP,
+			Value: ip.String(),
+		})
+	}
+
+	return Normalize(sans)
+}
+
+// Normalize returns the set of all unique ACME identifiers in the
+// input after all of them are lowercased. The returned identifier values will
+// be in their lowercased form and sorted alphabetically by value.
+func Normalize(idents []ACMEIdentifier) []ACMEIdentifier {
+	for i := range idents {
+		idents[i].Value = strings.ToLower(idents[i].Value)
+	}
+
+	sort.Slice(idents, func(i, j int) bool {
```

Comment:
The newer `slices.SortFunc` is [preferred over](https://pkg.go.dev/sort#Slice) `sort.Slice`, and in fact most of the functions in the old `sort` package have been reimplemented in terms of the more efficient functions from the `slices` package.

Also, using string formatting to perform the comparison is probably significantly less efficient than using multiple conditionals (i.e. if their types are the same just compare their values, otherwise return the dns first).